### PR TITLE
addrman: Fix format string in deserialize error

### PR DESCRIPTION
**Pull request by maflcko**
The format string is evaluated differently on modern compilers (clang 10 and later, as well as gcc 10 and later).

Work around the behaviour change in compilers by pinning the underlying type of the format arguments.

Can be tested by observing a failing test when running against master compiled with clang 10 or gcc 10 (or later).

### CODE REVIEW

**Comment by practicalswift** [comment]
Context for other reviewers: https://github.com/c42f/tinyformat/issues/82

**Comment by MarcoFalke** [comment]
> Context for other reviewers: c42f/tinyformat#82

I think the changes here make sense even without the issue (and whether it is a bug or not).

File: test/functional/feature_addrman.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        write_addrman(peers_dat)
+        with self.nodes[0].assert_debug_log(["Loaded 0 addresses from peers.dat"]):
+            self.start_node(0, extra_args=["-checkaddrman=1"])
+        assert_equal(self.nodes[0].getnodeaddresses(), [])
+
+        self.log.info("Check that addrman from future cannot be read")
+        self.stop_node(0)
+        write_addrman(peers_dat, lowest_compatible=111)
+        with self.nodes[0].assert_debug_log([
+                f'ERROR: DeserializeDB: Deserialize or I/O error - Unsupported format of addrman database: 1. It is compatible with formats >=111, but the maximum supported by this version of {self.config["environment"]["PACKAGE_NAME"]} is 3.',
```
Comment (jonatack) [inline test/functional/feature_addrman.py:58]: nit, line length
```diff
-        f'ERROR: DeserializeDB: Deserialize or I/O error - Unsupported format of addrman database: 1. It is compatible with formats >=111, but the maximum supported by this version of {self.config["environment"]["PACKAGE_NAME"]} is 3.',
+       "ERROR: DeserializeDB: Deserialize or I/O error - Unsupported format of "
+        "addrman database: 1. It is compatible with formats >=111, but the maximum "
+        f"supported by this version of {self.config['environment']['PACKAGE_NAME']} is 3.",
```

**Review by jonatack — COMMENTED** [review]
ACK fab0b55cf060c2b14fae5cee13f0a2dcaebde892 verified the test fails on master as expected only at line 61 (assertion fixed by the code change); the last two test additions pass as expected

```
2021-09-06T13:36:53.543000Z TestFramework (INFO): Check that addrman from future cannot be read
2021-09-06T13:36:55.715000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 131, in main
    self.run_test()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/feature_addrman.py", line 61, in run_test
    self.start_node(0)
  File "/usr/lib/python3.9/contextlib.py", line 126, in __exit__
    next(self.gen)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 400, in assert_debug_log
    self._raise_assertion_error('Expected messages "{}" does not partially match log:\n\n{}\n\n'.format(str(expected_msgs), print_log))
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 166, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] Expected messages "['ERROR: DeserializeDB: Deserialize or I/O error - Unsupported format of addrman database: 1. It is compatible with formats >=111, but the maximum supported by this version of Bitcoin Core is 3.', 'Recreating peers.dat']" does not partially match log:
...
 - 2021-09-06T13:36:53.840640Z [init] [util/system.h:51] [error] ERROR: DeserializeDB: Deserialize or I/O error - Unsupported format of addrman database: \x01. It is compatible with formats >=111, but the maximum supported by this version of Bitcoin Core is 3.: iostream error

```

File: test/functional/feature_addrman.py (hunk 1 above)
Comment (MarcoFalke) [reply test/functional/feature_addrman.py:58]: Thanks, will fix on next force push, if there is one

File: src/addrman.cpp [hunk 2]
```diff
@@ -242,9 +242,9 @@ void CAddrMan::Unserialize(Stream& s_)
     const uint8_t lowest_compatible = compat - INCOMPATIBILITY_BASE;
     if (lowest_compatible > FILE_FORMAT) {
         throw std::ios_base::failure(strprintf(
-                    "Unsupported format of addrman database: %u. It is compatible with formats >=%u, "
-                    "but the maximum supported by this version of %s is %u.",
-                    format, lowest_compatible, PACKAGE_NAME, static_cast<uint8_t>(FILE_FORMAT)));
+            "Unsupported format of addrman database: %u. It is compatible with formats >=%u, "
+            "but the maximum supported by this version of %s is %u.",
+            uint8_t{format}, uint8_t{lowest_compatible}, PACKAGE_NAME, uint8_t{FILE_FORMAT}));
```
Comment (mzumsande) [inline src/addrman.cpp:247]: nit: also pinning `lowest_compatible` seems unnecessary.

**Review by mzumsande — COMMENTED** [review]
ACK fab0b55cf060c2b14fae5cee13f0a2dcaebde892
I verified with clang 10 that this corrects the format. 
facce4ca44bc206b7656e297a7fa5dfb83a01012 looks ok but unrelated.

File: src/addrman.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/addrman.cpp:247]: Correct. I did that for consistency.

**Comment by MarcoFalke** [comment]
>  facce4c looks ok but unrelated.

Right. Didn't seem worth it to open a separate pull for this "trivial style" fix. I found it while working on this pull, so at least it is loosely related :sweat_smile:

**Pull request merged**